File: chap16_endocrine/lacker.m

%    -------------------------------------------------------------------
%
% Solve the Lacker model for the control of ovulation number. The model is
% solved in the transformed coordinates (gamma rather than xi) because it is
% much easier that way, as this avoids the problem of  solutions going to infinity in
% finite time, which is a real pain
%
%     For Chapter 16, Section 16.3.2 of
%     Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
%
%     Written by James Keener and James Sneyd.
%
%    -------------------------------------------------------------------

function lacker

clear all; close all; clc;

set(0,                           ...
   'defaultaxesfontsize', 20,   ...
   'defaultaxeslinewidth', 1.2, ...
   'defaultlinelinewidth', 2.0);
par.n = 10;
par.M1 = 4;
par.M2 = 20;

fol0 = sort(0.1*rand(1,par.n),'descend'); % They must be ordered.
xi10 = fol0(1);
gam0 = fol0/fol0(1);  % initial conditions for the gammas.
tspan = linspace(0,2,1000);
options = odeset('RelTol',1e-7,'AbsTol',1e-7);

% First solve for the gammas, xi1 and t. Easier to do it all together.
[tau,y]=ode15s(@(tau,y)rhs(tau,y,par),tspan,[gam0,xi10,0],options);
gam = y(:,1:par.n);
xi1 = y(:,par.n+1);
t = y(:,par.n+2);
figure(1)
plot(tau,gam)
xlabel('\tau')
ylabel('\gamma_i')

% Now reconstruct and plot the xi
xi = gam.*xi1;
ov_index = find(xi(:,1)>5,1);
t_ov = t(ov_index)
ov_num = sum(gam(end,:)>0.9)

figure(2)
plot(t,xi)
ylim([0 5])
xlabel('t')
ylabel('\xi_i')


end
%%
function out=rhs(tau,y,par)
gam = y(1:par.n);
xi1 = y(par.n+1);
GG = sum(gam);
gam_deriv = gam.*(1-gam).*(par.M1*par.M2*(1+gam) - GG*(par.M1+par.M2));
xi_deriv = (1/xi1)*(1 - xi1^2*(GG-par.M1)*(GG-par.M2));
time_deriv = 1/(xi1^2);
out = [gam_deriv',xi_deriv,time_deriv]';
end

File: chap16_endocrine/lacker.py

#     -------------------------------------------------------------------
# 
#  Solve the Lacker model for the control of ovulation number. The model is
#  solved in the transformed coordinates (gamma rather than xi) because it is
#  much easier that way, as this avoids the problem of  solutions going to infinity in
#  finite time, which is a real pain
# 
#      For Chapter 16, Section 16.3.2 of
#      Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#      Written by James Keener and James Sneyd.
# 
#     -------------------------------------------------------------------

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp


# Define the right-hand side of the ODE system
def rhs(t, y):
    gam = y[:n]
    xi1 = y[n]
    GG = np.sum(gam)
    gam_deriv = gam * (1 - gam) * (M1 * M2 * (1 + gam) - GG * (M1 + M2))
    xi_deriv = (1 / xi1) * (1 - xi1 ** 2 * (GG - M1) * (GG - M2))
    time_deriv = 1 / (xi1 ** 2)
    return np.concatenate((gam_deriv, [xi_deriv, time_deriv]))


# Define parameters
n = 10
M1 = 4
M2 = 20

# Initial conditions
fol0 = np.sort(0.1 * np.random.rand(n))[::-1]  # They must be ordered.
xi10 = fol0[0]
gam0 = fol0 / fol0[0]  # Initial conditions for the gammas.
y0 = np.concatenate((gam0, [xi10, 0]))

# Define the time span
tspan = [0, 2]
tout=np.linspace(tspan[0], tspan[1], 1000)
# Solve the ODE system
sol = solve_ivp(rhs, tspan, y0, method='LSODA', t_eval=tout)

# Extract the results
gam = sol.y[:n]
xi1 = sol.y[n]
t = sol.y[n + 1]

# Plot the results
plt.figure(1)
plt.plot(t, gam.T)
plt.xlabel('tau')
plt.ylabel('gamma_i')

ov_index = np.argmax(xi1 > 5)
t_ov = t[ov_index]
ov_num = np.sum(gam[:, -1] > 0.9)
print("Ovulation time:", t_ov)
print("Number of ovulations:", ov_num)

plt.figure(2)
plt.plot(t, xi1 * gam.T)
plt.xlabel('t')
plt.ylabel('xi_i')
plt.ylim([0, 5])

plt.show()
